### PR TITLE
ILM wait for snapshot creation to complete

### PR DESCRIPTION
**Pull request by andreidan**
As the snapshot is not taken on the calling thread even though the request
wait_for_completion parameter is true, we can set the flag to true and make
use of the response SnapshotInfo to verify the snapshot was created successfully
or go back to the CleanSnapshotStep and retry creating the snapshot otherwise.
This branching is executed using a new step that wraps an AsyncActionStep
(keeping the wrapped's step key()) and moving to a different next step when
the wrapped action reports an incomplete response.

### CODE REVIEW

**Comment by andreidan** [comment]
@elasticmachine run elasticsearch-ci/2

**Comment by andreidan** [comment]
Integration tests flakiness might be encountered until https://github.com/elastic/elasticsearch/pull/54433 is merged

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/AsyncActionBranchingStep.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+            public void onFailure(Exception e) {
+                listener.onFailure(e);
+            }
+        });
+    }
+
+    @Override
+    public final StepKey getNextStepKey() {
+        if (onResponseResult.get() == null) {
+            throw new IllegalStateException("Cannot call getNextStepKey before performAction");
```
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/AsyncActionBranchingStep.java:64]: Super minor nit
```suggestion
            throw new IllegalStateException("cannot call getNextStepKey before performAction");
```

**Review by dakrone — APPROVED** [review]
LGTM, I added one super minor nit comment, thanks Andrei!

**Comment by andreidan** [comment]
@elasticmachine update branch

**Pull request merged**